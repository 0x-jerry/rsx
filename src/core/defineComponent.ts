import { isFn } from '@0x-jerry/utils'
import type { Merge } from 'type-fest'
import type { ToMaybeRef } from '.'
import type { DefineProps } from './props'

export interface PropOption<T = any> {
  type?: T
  default?: T | (() => T)
}

export type PropsType<T = any> = Record<string, T | PropOption<T>>

export interface ComponentOption<Props extends PropsType = {}> {
  props?: Props
  toRefs?: boolean
}

interface CommonProps {
  class?: any
  [key: string]: any
}

export type FunctionalComponent<P extends PropsType = {}> = (
  props: Merge<P, CommonProps>,
  children?: any[],
) => JSX.Element

/**
 * without option
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * with option
 * @param opt
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P>,
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * implement
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P> | FunctionalComponent,
  impl?: FunctionalComponent,
): FunctionalComponent {
  if (isFn(opt)) return opt

  // todo check props type
  return impl!
}

export const dc = defineComponent
