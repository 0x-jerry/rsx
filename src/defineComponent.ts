import type { EmptyObject } from '@0x-jerry/utils'
import type { DefineProps } from './props'
import type { IsStartWithCapitalizedLetter, MaybeRef } from './types'

export type PropsType<T = any> = Record<string, T>

type ToPropRefs<T extends {}> = {
  /**
   * Start with capitalized letter means this prop is not used as reactive prop
   */
  [key in keyof T]: key extends IsStartWithCapitalizedLetter<key>
    ? T[key]
    : MaybeRef<T[key]>
}

/**
 * @private
 */
export type FunctionalComponent<P extends PropsType = any> = (
  props: P,
  children?: any[],
) => unknown

type _NoInferProps<P extends EmptyObject> = DefineProps<P> & Record<string, any>

export type ExposedFunctionalComponent<P extends PropsType = any> =
  FunctionalComponent<ToPropRefs<P>>

/**
 * Auto calculate props from generic type P
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: NoInfer<FunctionalComponent<_NoInferProps<P>>>,
): ExposedFunctionalComponent<P>

/**
 * Infer props from impl parameter
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<P>,
): ExposedFunctionalComponent<P>

/**
 * implement
 */
export function defineComponent(impl?: any): any {
  // todo check props type
  return impl
}

export const dc = defineComponent
