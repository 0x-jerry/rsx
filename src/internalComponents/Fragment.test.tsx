import { dc } from '../defineComponent'
import { contextToJson, defineComponentName, mountTestApp } from '../test'

describe('Fragment', () => {
  it('non-reactivity data', async () => {
    const App = dc(() => {
      return (
        <>
          <div class="item">1</div>
          <div class="item">2</div>
        </>
      )
    })

    const el = mountTestApp(App)

    const contents = el
      .querySelectorAll('.item')
      .values()
      .map((item) => item.textContent)
      .toArray()

    expect(contents).eql(['1', '2'])
  })
})

describe('fragment context tree', () => {
  it('static value', () => {
    const A = dc((_, children) => (
      <div class="A">
        <span>a</span>
        {children}
      </div>
    ))

    defineComponentName(A, 'A')

    const App = dc(() => (
      <>
        1<A></A>2
      </>
    ))

    defineComponentName(App, 'App')

    const root = mountTestApp(App)

    expect(root.outerHTML).toMatchSnapshot()

    const ctxTree = contextToJson(root._)

    expect(ctxTree).toMatchSnapshot()
  })
})
