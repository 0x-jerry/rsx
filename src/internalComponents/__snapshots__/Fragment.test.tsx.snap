// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fragment context tree > static value > ctx tree 1`] = `
{
  "children": [
    {
      "children": [
        {
          "children": [],
          "name": "A",
        },
      ],
      "name": "Fragment",
    },
  ],
  "name": "Root",
}
`;

exports[`fragment context tree > static value > html 1`] = `
<div
  id="app"
>
  1
  <div
    class="A"
  >
    <span>
      a
    </span>
  </div>
  2
  <!--fragment-->
</div>
`;
